### PR TITLE
Add shard partition peeking

### PR DESCRIPTION
**Pull request by tonyhb**
This implements shard partition peeking.  For workers which lease shards, we randomly choose a leased shard to peek (fetch functions), with 5% chance of still working from the global partition.

## Type of change (choose one)
- [x] Chore (refactors, upgrades, etc.)
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Security fix (non-breaking change that fixes a potential vulnerability)
- [ ] Docs
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)

## Checklist
- [ ] I've linked any associated issues to this PR.
- [ ] I've tested my own changes.

*[Check our Pull Request Guidelines](https://github.com/inngest/inngest/blob/main/docs/PULL_REQUEST_GUIDELINES.md)*

### CODE REVIEW

File: pkg/execution/state/redis_state/queue_processor.go [hunk 1]
```diff
@@ -278,6 +296,9 @@
 						q.logger.Error().Err(err).Msg("error scanning and leasing shards")
 						return
 					}
+					if retry {
+						<-time.After(time.Duration(rand.Intn(20)) * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (tonyhb) [reply pkg/execution/state/redis_state/queue_processor.go:300]: This is fine.  We don't need security, just semi-randomness.

File: pkg/execution/state/redis_state/queue_processor.go [hunk 2]
```diff
@@ -315,15 +338,23 @@ func (q *queue) scanShards(ctx context.Context) (retry bool, err error) {
...
 
 	for _, shard := range shards {
 		leaseID, err := q.leaseShard(ctx, shard, ShardLeaseTime, len(shard.Leases))
 		if err == nil {
+			// go q.counter(ctx, "queue_shard_lease_success_total", 1, map[string]any{
+			// 	"shard_name": shard.Name,
+			// })
 			q.addLeasedShard(ctx, shard, *leaseID)
 		}
+		// go q.counter(ctx, "queue_shard_lease_conflict_total", 1, map[string]any{
```
Comment (tonyhb) [inline pkg/execution/state/redis_state/queue_processor.go:354]: We'll add these after once the gauges are reporting in a followup PR.

File: pkg/execution/state/redis_state/lua/queue/enqueue.lua [hunk 3]
```diff
@@ -56,8 +56,13 @@ if shard ~= "" and shard ~= "null" then
     -- NOTE: We do not want to overwrite the shard leases, so here
     -- we fetch the shard item, set the lease values in the passed in shard
     -- item, then write the updated value.
-    --
-    -- TODO: LEASES ???
+    local existingShard = redis.call("HGET", shardMapKey, shardName)
+    if existingShard ~= nil and existingShard ~= false then
+	local updatedShard = cjson.decode(shard)
+	existingShard = cjson.decode(existingShard)
```
Comment (darwin67) [inline pkg/execution/state/redis_state/lua/queue/enqueue.lua:62]: nit: indentation is weird

File: pkg/execution/state/redis_state/queue.go [hunk 4]
```diff
@@ -437,6 +446,8 @@ type queue struct {
 
 	// metrics allows reporting of metrics
 	scope tally.Scope
+	meter metric.Meter
```
Comment (darwin67) [inline pkg/execution/state/redis_state/queue.go:449]: meters are global singletons, so as long as you initialize them like this, you shouldn't need to assign it.
https://github.com/inngest/monorepo/blob/develop/cmd/all-in-one/executor.go#L79-L90

File: pkg/execution/state/redis_state/queue_processor.go [hunk 5]
```diff
@@ -1066,6 +1127,61 @@ func (q *queue) isScavenger() bool {
...
+		}
+	}()
+
+	// Report gauges to otel.
+	_, _ = q.meter.Int64ObservableGauge(
+		"queue_shards_count",
+		metric.WithDescription("Number of shards in the queue"),
+		metric.WithInt64Callback(func(ctx context.Context, o metric.Int64Observer) error {
+			o.Observe(int64(len(shards)))
+			return err
```
Comment (darwin67) [inline pkg/execution/state/redis_state/queue_processor.go:1156]: where does this `err` come from?
same for the other gauges.

File: pkg/execution/state/redis_state/queue_processor.go [hunk 6]
```diff
@@ -1066,6 +1127,61 @@ func (q *queue) isScavenger() bool {
...
+			case <-tick.C:
+				// Reload shards.
+				shards, err = q.getShards(ctx)
+			}
+		}
+	}()
+
+	// Report gauges to otel.
+	_, _ = q.meter.Int64ObservableGauge(
+		"queue_shards_count",
```
Comment (darwin67) [inline pkg/execution/state/redis_state/queue_processor.go:1152]: let's add a prefix of `inngest` to all metric names so it's easier to identify.

File: pkg/execution/state/redis_state/queue_processor.go [hunk 7]
```diff
@@ -315,15 +338,23 @@ func (q *queue) scanShards(ctx context.Context) (retry bool, err error) {
...
 	}
+
 	if len(shards) == 0 {
 		return
 	}
 
 	for _, shard := range shards {
 		leaseID, err := q.leaseShard(ctx, shard, ShardLeaseTime, len(shard.Leases))
 		if err == nil {
+			// go q.counter(ctx, "queue_shard_lease_success_total", 1, map[string]any{
```
Comment (darwin67) [inline pkg/execution/state/redis_state/queue_processor.go:349]: no longer need this comment?

File: pkg/execution/state/redis_state/queue_processor.go (hunk 5 above)
Comment (tonyhb) [reply pkg/execution/state/redis_state/queue_processor.go:1156]: The loading above.

File: pkg/execution/state/redis_state/queue.go (hunk 4 above)
Comment (tonyhb) [reply pkg/execution/state/redis_state/queue.go:449]: Hmm, what would you prefer this be?

We need some way of creating singletons in this package, though, so I'd prefer to be passing down a meter here.  The queue doesn't have access to any global singletons and should declare what it needs.

File: pkg/execution/state/redis_state/lua/queue/enqueue.lua (hunk 3 above)
Comment (tonyhb) [reply pkg/execution/state/redis_state/lua/queue/enqueue.lua]: Fixed, thanks!

**Review by darwin67 — APPROVED** [review]

**Pull request merged**